Creating a CheckAndMutateBuilder

sduskis opened the pull request:
There was logic duplication in AbstractBigtableTable and BigtableAsyncTable.  The new CheckAndMutateUtil RequestBuilder combines the logic for both implementations.
Also, this should fix #1704. When value is null, and the CompareOp was not '==' or '!=' would do absolutely nothing, when it should have been an "non-existence" check.

igorbernstein2 commented:
Can we add an integration test that can run against the minicluster & bigtable verifying this behavior?

sduskis commented:
Sorry about the oversight.  I had a test ready to go in another branch.

igorbernstein2 commented on bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/CheckAndMutateUtil.java:
```diff
@@ -71,67 +70,129 @@ public static boolean wasMutationApplied(
+      return this;
+    }
+
+    public RequestBuilder ifEquals(CompareOp compareOp, byte[] value) {
```
I don't think this is used

sduskis replied:
Thanks for the catch.

igorbernstein2 commented on bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/CheckAndMutateUtil.java:
```diff
@@ -71,67 +70,129 @@ public static boolean wasMutationApplied(
+      return ifMatches(CompareOp.EQUAL, value);
+    }
+
+    public RequestBuilder ifMatches(CompareOp compareOp, byte[] value) {
```
I think it would be cleaner to stick to a builder style and have `setOperator` & `setOperand` or go the other way make this a DSL and add the qualifier as a parameter

sduskis replied:
HBase uses this approach, and I think I'd like to keep it for now.   I'd like to think through what a more powerful `CheckAndMutateRequest` builder would look like.  

The `with(Put|Delete|Mutation)` API allows for more than the single mutation, which is great.  It would be awesome to be able to add more conditions as well; something like:

```
builder
   .ifQualifierMatches(q1, ...)
   .ifQualifierNotExists(q2)
   .ifCondition(randomFilter)
`

igorbernstein2 commented on bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/CheckAndMutateUtil.java:
```diff
@@ -71,67 +70,129 @@ public static boolean wasMutationApplied(
+      return this;
+    }
+
+    public CheckAndMutateRowRequest buildForPut(Put put) throws DoNotRetryIOException {
```
It might be better to split the 2 responsibilities:
`addPut` and `build()`

sduskis replied:
ack

sduskis commented on bigtable-hbase-parent/bigtable-hbase-integration-tests-common/src/main/java/com/google/cloud/bigtable/hbase/AbstractTestCheckAndMutate.java:
```diff
@@ -366,6 +370,54 @@ public void testCompareOps() throws Exception {
+        CompareOp.GREATER_OR_EQUAL, new byte[]{Byte.MIN_VALUE}, someRandomPut);
+    Assert.assertFalse("> MIN_BYTES should fail", success);
+
+    if (sharedTestEnv.isBigtable()) {
```
ok

sduskis commented:
I addressed the comments.  As per the comment on the outdated code, I would like to keep `ifMatches` as it is to reduce the chances of problematic situations where a user doesn't set both `compareOp` and an explicit value selection.

sduskis commented:
For now, I'm going to ask that we stay consistent with HBase 2.*'s API, at least for now.  Is that ok?
